providers/universal-provider/src/UniversalProvider.ts

```diff
@@ -0,0 +1,267 @@
+  }
+
+  public once(event: string, listener: any): void {
+    this.events.once(event, listener);
+  }
+
+  public removeListener(event: string, listener: any): void {
+    this.events.removeListener(event, listener);
+  }
+
+  public off(event: string, listener: any): void {
+    this.events.off(event, listener);
+  }
+
+  get isWalletConnect() {
+    return true;
+  }
+
+  public async pair(pairingTopic: string | undefined): Promise<SessionTypes.Struct> {
+    const { uri, approval } = await this.client.connect({
+      pairingTopic,
+      requiredNamespaces: this.namespaces,
+    });
+
+    if (uri) {
+      this.uri = uri;
+      this.events.emit("display_uri", uri);
+    }
+
+    this.session = await approval();
+    this.onSessionUpdate();
+    return this.session;
+  }
+
+  public async setDefaultChain(chain: string, rpcUrl?: string | undefined) {
+    const [namespace, chainId] = this.validateChain(chain);
+    this.getProvider(namespace).setDefaultChain(chainId, rpcUrl);
+  }
+
+  // ---------- Private ----------------------------------------------- //
+
+  private async checkStorage() {
+    this.namespaces = (await this.client.core.storage.getItem(
+      `${STORAGE}/namespaces`,
+    )) as NamespaceConfig;
+    if (this.namespaces) {
+      this.createProviders();
+    }
+
+    if (this.client.session.length) {
+      const lastKeyIndex = this.client.session.keys.length - 1;
+      this.session = this.client.session.get(this.client.session.keys[lastKeyIndex]);
+    }
+  }
+
+  private async initialize() {
+    this.logger.trace(`Initialized`);
+    await this.createClient();
+    this.checkStorage();
+    this.registerEventListeners();
+  }
+
+  private async createClient() {
+    this.client =
+      this.providerOpts.client ||
+      (await SignClient.init({
+        logger: this.providerOpts.logger || LOGGER,
+        relayUrl: this.providerOpts.relayUrl || RELAY_URL,
+        projectId: this.providerOpts.projectId,
+        metadata: this.providerOpts.metadata, // fetch metadata automatically if not provided?
+      }));
+
+    this.logger.trace(`SignClient Initialized`);
+  }
+
+  private createProviders(): void {
+    if (!this.client) {
+      throw new Error("Client not initialized");
+    }
+
+    Object.keys(this.namespaces).forEach((namespace) => {
+      switch (namespace) {
+        case "eip155":
+          this.rpcProviders[namespace] = new Eip155Provider({
+            // @ts-ignore
+            client: this.client,
+            namespace: this.namespaces[namespace],
+            events: this.events,
+          });
+          break;
+        case "solana":
+          //TODO:
+          break;
+        case "cosmos":
+          //TODO:
+          break;
+        case "polkadot":
+          //TODO:
+          break;
+      }
+    });
+  }
+
+  private registerEventListeners(): void {
+    if (typeof this.client === "undefined") {
+      throw new Error("WalletConnect is not initialized");
```

Comment:
Consistency nitpick:

```suggestion
      throw new Error("Sign Client is not initialized");
```